File: THCS/2022_2023/HCM_2223/Bai2/Bai2.py
from itertools import product

def find_all_strings(n, m, base_string, replacements):
    # Lấy tất cả tổ hợp các ký tự thay thế cho các dấu #
    combinations = product(*replacements)  # Tạo ra tất cả tổ hợp thay thế cho các dấu #

    result_list = []

    for combination in combinations:
        result = []
        hash_count = 0  # Đếm số lượng dấu #
        for i in range(n):
            if base_string[i] == '#':
                # Thay thế dấu # bằng ký tự trong tổ hợp hiện tại
                result.append(combination[hash_count])
                hash_count += 1
            else:
                # Giữ nguyên các ký tự không phải dấu #
                result.append(base_string[i])
        # Thêm chuỗi tạo ra vào danh sách kết quả
        result_list.append(''.join(result))

    # Sắp xếp tất cả các chuỗi theo thứ tự từ điển
    result_list.sort()
    return result_list

# Đọc dữ liệu đầu vào
n, m, k, x = map(int, input().split())
base_string = input().strip()

replacements = []
for _ in range(m):
    replacements.append(sorted(input().strip()))  # Lưu các ký tự thay thế sau khi sắp xếp theo thứ tự từ điển

# Tìm tất cả các chuỗi có thể thay thế
all_strings = find_all_strings(n, m, base_string, replacements)

# Xuất ra tất cả các chuỗi
for i, string in enumerate(all_strings, 1):
    print(f"{i}. {string}")

File: THCS/2022_2023/HCM_2223/Bai2/Bai2.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

// Hàm đệ quy để tạo ra tất cả các chuỗi
void generateStrings(int idx, const string &base, vector<string> &results, const vector<vector<char>> &replacements, string current) {
    if (idx == base.size()) {
        results.push_back(current); // Thêm chuỗi hiện tại vào kết quả
        return;
    }
    
    if (base[idx] == '#') {
        for (char ch : replacements[idx]) {
            generateStrings(idx + 1, base, results, replacements, current + ch); // Thay thế #
        }
    } else {
        generateStrings(idx + 1, base, results, replacements, current + base[idx]); // Giữ nguyên ký tự
    }
}

int main() {
    int n, m, k, x;
    cin >> n >> m >> k >> x;
    cin.ignore(); // Đọc dòng mới sau khi đọc các số nguyên

    string base_string;
    getline(cin, base_string); // Đọc chuỗi n ký tự

    vector<vector<char>> replacements(m); // Lưu trữ các ký tự thay thế cho mỗi #
    
    for (int i = 0; i < m; ++i) {
        string replacement;
        getline(cin, replacement);
        replacements[i] = vector<char>(replacement.begin(), replacement.end());
        sort(replacements[i].begin(), replacements[i].end()); // Sắp xếp ký tự thay thế
    }

    vector<string> results; // Danh sách để lưu trữ các chuỗi kết quả
    generateStrings(0, base_string, results, replacements, ""); // Gọi hàm đệ quy

    // Sắp xếp danh sách các chuỗi theo thứ tự từ điển
    sort(results.begin(), results.end());

    // In ra tối đa 8 chuỗi
    for (int i = 0; i < min(8, (int)results.size()); ++i) {
        cout << i + 1 << ". " << results[i] << endl; // In ra các chuỗi theo thứ tự
    }

    return 0;
}
